### ISSUE TITLE
Unable to get client to authenticate

### ISSUE THREAD

**Issue by steveellis** [onovotny/SignService#59]
I have built and deployed the server and am able to create signing accounts and assign certificates to them but I'm unable to get the client to authenticate to AD when I run the client locally on my machine.

Here's my appsettings.json file for the local sign client:

```javascript
{
  "SignClient": {
    "AzureAd": {\n
      "AADInstance": "https://login.microsoftonline.com/",
       "ClientId": "<my sign client AD client id>",
      "TenantId": "<my AD tenant id/directory>",
    },
    "Service": {
      "Url": "<The endpoint of my sign service as an azure web app>",
      "ResourceId": "<The resource ID as shown in the AD server app>"
    }
  }
}
```

I'm getting AADSTS70002: Error validating credentials. AADSTS50126: Invalid username or password I tried to create a simplified powershell script for the client.

```powershell
$currentDirectory = split-path $MyInvocation.MyCommand.Definition

# Tried to use the ClientSecret returned by the server install utility.
# That seems wrong though since this is the client. Tried both just in case.
$secret = "<a secret generated from the client AD app>"

# If it doesn't include the stuff after the @ sign then it will say 'user not found'. So I think my username is ok.
$user = "<a sign server username as created in the server app>@<mydomain>.onmicrosoft.com"

$appSettings = "$currentDirectory\appsettings.json"
$toSign = "$currentDirectory\FileToSign.txt";

& SignClient sign -c $appSettings -r $user -s $secret -n 'SignTest0' -i $toSign
```

Anyone have any ideas about what I can try?

What I've tried so far:

* Although the docs don't seem to mention it, the sign client app seems to need its own secret (does it?) which as far as I can tell can only be generated in the portal by going to Azure Active Directory > App Registrations (Preview).

* The server secret that is used to authenticate to Key Vault doesn't work for the client for me anyway. My server build seems to be configured correctly. I can log into it. I can create users and add certificates.

* I've tried fiddling with the ADDInstance property in the client appsettings.json. Thought it should perhaps be `https://login.microsoftonline.com/<my tenant guid>` but that doesn't work. Although that is the endpoint I needed in order to get the `InstallUtility` to run.

* There are a few different client ids that are generated. One I believe is for the server and the other is for the sign client. The `InstallUtility` returns two, both of which are labeled ClientId which is a little confusing. I've tried using both values returned there in the client appsettings.json and also have learned how to find the ClientId for the client AD app in AD in the portal. But plugging that value doesn't help things.

* It wasn't immediately clear how the sign client userid should be written. But I'm pretty sure it is such like `<a sign server username as created in the server app>@<mydomain>.onmicrosoft.com` because otherwise I get a user not found error. So it seems AD is finding my user. Which would imply the secret is wrong.

* So I've regenerated the AD client app secret several times in the portal just to see if that would help but sadly it does not.

**Comment by onovotny** [comment]
Hi @steveellis, sorry for troubles! 

The user id for the sign client should be the full user name `user@domain`. For the app, it needs the password for the user, it does not have its own secret. The app supports a clientid/secret *or* user/password mechanism, as an earlier version of the sign service worked with service principals instead.

Are you using the auto-generated password for the sign client? If have forgotten it, the password can be reset in the admin UI.

**Comment by steveellis** [comment]
Thanks Oren for your response. I thought I had tested using the user password generated from the server sign user create process for the sign client secret, but it looks like it wasn't working because I neglected to put the password in single quotes in my PowerShell script. Powershell was parsing the string and hitting a "$" char in the password. Now that I've fixed that, the $secret parameter the user name are in the format I was using works just fine.

Hopefully this will help someone else. The sign client secret should be the password for the user. Don't be distracted by all the other "client secrets" -- they're for other stuff!
